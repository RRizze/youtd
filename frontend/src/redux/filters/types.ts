import { ActionCreator } from "redux";

export interface SelectFilterState {
  select: boolean;
  not: boolean;
}

export interface RangeFilterState {
  min: number;
  max: number;
}

export interface Rarity {
  all: boolean;
  common: SelectFilterState;
  uncommon: SelectFilterState;
  rare: SelectFilterState;
  unique: SelectFilterState;
}

export interface Elements {
  all: boolean;
  nature: SelectFilterState;
  fire: SelectFilterState;
  ice: SelectFilterState;
  astral: SelectFilterState;
  darkness: SelectFilterState;
  storm: SelectFilterState;
  iron: SelectFilterState;
}

export interface AttackTypes {
  all: boolean;
  elemental: SelectFilterState;
  physical: SelectFilterState;
  energy: SelectFilterState;
  decay: SelectFilterState;
  magic: SelectFilterState;
  essence: SelectFilterState;
}

export interface SelectFilters {
  rarity: Rarity;
  elements: Elements;
  attack_types: AttackTypes;
}

export type SelectFiltersUnion = Rarity | Elements | AttackTypes;

export interface RangeFilters {
  damage: RangeFilterState;
  range: RangeFilterState;
  attack_cd: RangeFilterState;
  element_lvl_up: RangeFilterState;
  wave_lvl_up: RangeFilterState;
  goldcost: RangeFilterState;
}

export interface FiltersState {
  count_rows: number;
  offset_rows: number;
  select: SelectFilters;
  ranges: RangeFilters;
};
