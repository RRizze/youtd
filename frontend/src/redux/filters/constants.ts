export const SET_SELECT_FILTER = "SET_SELECT_FILTER";
export const RESET_SELECT_FILTER = "RESET_SELECT_FILTER";
export const SET_RANGE_VALUE = "SET_RANGE_VALUE";
export const RESET_COUNT_ROWS = "RESET_COUNT_ROWS";
export const SET_COUNT_ROWS = "SET_COUNT_ROWS";
export const SET_NOT_FILTER_SELECT = "SET_NOT_FILTER_SELECT";
