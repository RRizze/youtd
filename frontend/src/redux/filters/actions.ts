import * as constants from "./constants";
import { SelectFiltersUnion } from "./types";
import { createAction, ActionsUnion } from "../types";

export const setFilterAll = (
  filter: string,
  filters: SelectFiltersUnion,
  not?: string
) => {
  let newFilters;
  if (not !== "not") {
    newFilters = {
      ...filters,
      [filter]: {
        select: !filters[filter].select,
        not: false,
      },
    };
  } else {
    newFilters = {
      ...filters,
      [filter]: {
        select: false,
        not: !filters[filter].not,
      },
    };
  }

  for (const filter in newFilters) {
    if (
      (newFilters[filter]["select"] || newFilters[filter]["not"])
      && newFilters[filter] !== "all"
    ) {
      return false;
    }
  }
  return true;
};

export const setNotSelectFilter = (
  filter: string,
  filterType: string,
  filters: SelectFiltersUnion
) => createAction(constants.SET_NOT_FILTER_SELECT, {
  filter,
  filterType,
  all: setFilterAll(filter, filters, "not"),
});

export const setSelectFilter = (
  filter: string,
  filterType: string,
  filters: SelectFiltersUnion
) => createAction(constants.SET_SELECT_FILTER, {
  filter,
  filterType,
  all: setFilterAll(filter, filters),
});

export const resetSelectFilter = (filterType: string) => createAction(constants.RESET_SELECT_FILTER, {
  filterType,
});

export const setCountRows = () => createAction(constants.SET_COUNT_ROWS);
export const resetCountRows = () => createAction(constants.RESET_COUNT_ROWS);

export const setRangeFilter = (
  min: number,
  max: number,
  filterType: string
) => createAction(constants.SET_RANGE_VALUE, {
  min,
  max,
  filterType,
});

const Actions = {
  setSelectFilter,
  resetSelectFilter,
  setCountRows,
  resetCountRows,
  setNotSelectFilter,
  setRangeFilter,
};

export type Actions = ActionsUnion<typeof Actions>;
