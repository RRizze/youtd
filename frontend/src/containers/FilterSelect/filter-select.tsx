import React from "react";
import { connect } from "react-redux";
import "./filter-select-style.scss";
import {
  setSelectFilter,
  resetSelectFilter,
  setNotSelectFilter
} from "../../redux/filters/actions";
import { SelectFiltersUnion } from "../../redux/filters/types";

export interface Props {
  title: string;
  filterType: string;
  filters: SelectFiltersUnion;
  setSelectFilter: typeof setSelectFilter;
  resetSelectFilter: typeof resetSelectFilter;
  setNotSelectFilter: typeof setNotSelectFilter;
}

export const FilterSelect = ({
  title,
  filters,
  filterType,
  setSelectFilter,
  resetSelectFilter,
  setNotSelectFilter,
}: Props) => {
  const items = [];

  for (const filter in filters) {
    if (filter === "all") {
      items.push(
        <li className="filter-select-list__item" key={filter}>
          <span
            className={
              filters[filter]
                ? "filter-select-all--value active"
                : "filter-select-all--value"}
            onClick={() => resetSelectFilter(filterType)}>
            {filter}
          </span>
        </li>
      );
    } else {
      items.push(
        <li className="filter-select-list__item" key={filter}>
          <span
            className={
              filters[filter]["select"]
                ? "filter-select--value active"
                : "filter-select--value"}
            onClick={() => setSelectFilter(filter, filterType, filters)}
          >
            {filter}
          </span>
          <span
            className={
              filters[filter]["not"]
                ? "filter-select--not-value active"
                : "filter-select--not-value"}
            onClick={() => setNotSelectFilter(filter, filterType, filters)} >
            -
          </span>
        </li>
      );
    }
  }

  return (
    <div className="filter-select">
      <div className="filter-header">
        <h2 className="filter-title">{title}</h2>
      </div>
      <div className="filter-body">
        <ul className="filter-select-list">
          {items}
        </ul>
      </div>
    </div>
  );
};

export default connect(
  null,
  {
    setSelectFilter,
    resetSelectFilter,
    setNotSelectFilter,
  }
)(FilterSelect);
