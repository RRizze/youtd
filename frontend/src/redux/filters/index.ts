import * as constants from "./constants";
import { Actions } from "./actions";
import { FiltersState } from "./types";

export const initialState: FiltersState = {
  count_rows: 10,
  offset_rows: 0,
  select: {
    rarity: {
      all: true,
      common: {
        select: false,
        not: false,
      },
      uncommon: {
        select: false,
        not: false,
      },
      rare: {
        select: false,
        not: false,
      },
      unique: {
        select: false,
        not: false,
      },
    },
    elements: {
      all: true,
      nature: {
        select: false,
        not: false,
      },
      fire: {
        select: false,
        not: false,
      },
      ice: {
        select: false,
        not: false,
      },
      astral: {
        select: false,
        not: false,
      },
      darkness: {
        select: false,
        not: false,
      },
      storm: {
        select: false,
        not: false,
      },
      iron: {
        select: false,
        not: false,
      },
    },
    attack_types: {
      all: true,
      elemental: {
        select: false,
        not: false,
      },
      physical: {
        select: false,
        not: false,
      },
      energy: {
        select: false,
        not: false,
      },
      decay: {
        select: false,
        not: false,
      },
      magic: {
        select: false,
        not: false,
      },
      essence: {
        select: false,
        not: false,
      },
    },
  },
  ranges: {
    damage: {
      min: 1,
      max: 20000,
    },
    range: {
      min: 0,
      max: 10000,
    },
    attack_cd: {
      min: 0,
      max: 10,
    },
    element_lvl_up: {
      min: 0,
      max: 15,
    },
    wave_lvl_up: {
      min: 0,
      max: 90,
    },
    goldcost: {
      min: 0,
      max: 10000,
    },
  },
};

export default (state = initialState, action: Actions) => {
  switch (action.type) {
    case constants.RESET_COUNT_ROWS:
      return {
        ...state,
        offset_rows: 0,
      };
    case constants.SET_COUNT_ROWS:
      return {
        ...state,
        offset_rows: state.offset_rows + 10,
      };
    case constants.SET_SELECT_FILTER:
      return {
        ...state,
        offset_rows: 0,
        select: {
          ...state.select,
          [action.payload.filterType]: {
            ...state.select[action.payload.filterType],
            all: action.payload.all,
            [action.payload.filter]: {
              select: !state.select[action.payload.filterType][action.payload.filter].select,
              not: false,
            },
          },
        },
      };
    case constants.RESET_SELECT_FILTER:
      return {
        ...state,
        offset_rows: 0,
        select: {
          ...state.select,
          [action.payload.filterType]: {
            ...initialState.select[action.payload.filterType],
          },
        },
      };
    case constants.SET_RANGE_VALUE:
      return {
        ...state,
        offset_rows: 0,
        ranges: {
          ...state.ranges,
          [action.payload.filterType]: {
            min: action.payload.min,
            max: action.payload.max,
          },
        },
      };
    case constants.SET_NOT_FILTER_SELECT:
      return {
        ...state,
        offset_rows: 0,
        select: {
          ...state.select,
          [action.payload.filterType]: {
            ...state.select[action.payload.filterType],
            all: action.payload.all,
            [action.payload.filter]: {
              select: false,
              not: !state.select[action.payload.filterType][action.payload.filter].not,
            },
          },
        },
      };
    default:
      return state;
  }
};
