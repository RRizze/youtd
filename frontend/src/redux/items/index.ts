import * as constants from "./constants";
import { Actions } from "./actions";
import { State } from "./types";

const initialState: State = {
  list: [],
  isFetching: false,
  count: 0,
  error: false,
};

export default (state = initialState, action: Actions): object => {
  switch (action.type) {
    case constants.CLEAR_ITEMS_LIST:
      return {
        ...initialState,
      };
    case constants.LOAD_ITEMS_REQUEST:
      return {
        ...state,
        isFetching: true,
      };
    case constants.LOAD_ITEMS_SUCCESS:
      return {
        ...state,
        count: action.payload.count,
        isFetching: false,
        list: [
          ...state.list,
          ...action.payload.items,
        ],
      };
    case constants.LOAD_ITEMS_FAILURE:
      return {
        ...initialState,
        error: true,
      };
    default:
      return state;
  }
};
