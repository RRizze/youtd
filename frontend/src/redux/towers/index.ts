import * as actions from "./constants";
import { State } from "./types";
import { Actions } from "./actions";

export const initialState: State = {
  list: [],
  isFetching: false,
  count: 0,
  error: false,
};

export default (state = initialState, action: Actions) => {
  switch (action.type) {
    case actions.CLEAR_TOWERS_LIST:
      return {
        ...initialState,
      };
    case actions.LOAD_TOWERS_REQUEST:
      return {
        ...state,
        isFetching: true,
      };
    case actions.LOAD_TOWERS_SUCCESS:
      return {
        ...state,
        isFetching: false,
        count: action.payload.count,
        list: [
          ...state.list,
          ...action.payload.towers,
        ],
      };
    case actions.LOAD_TOWERS_FAILURE:
      return {
        ...initialState,
        error: true,
      };
    default:
      return state;
  }
};
