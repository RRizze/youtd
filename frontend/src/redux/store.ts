import { createStore, applyMiddleware, combineReducers } from "redux";
import thunk from "redux-thunk";
import towers from "./towers";
import filters from "./filters";
import items from "./items";
import { createLogger } from "redux-logger";

const middleware = [ thunk ];

if (process.env.NODE_ENV !== "production") {
  middleware.push(createLogger());
}

export default createStore(
  combineReducers({
    towers,
    filters,
    items,
  }),
  applyMiddleware(...middleware)
);
