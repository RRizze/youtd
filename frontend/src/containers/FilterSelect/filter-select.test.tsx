import React from "react";
import { mount } from "enzyme";
import renderer from "react-test-renderer";
import { FilterSelect } from "./filter-select";
import { initialState } from "../../redux/filters";
import {
  setSelectFilter,
  resetSelectFilter,
  setNotSelectFilter
} from "../../redux/filters/actions";

describe("Section tests", () => {
  it("render correctly", () => {
    const filters = {
      ...initialState.select.elements,
    };
    const props = {
      title: "elements",
      filters,
      filterType: "nature",
      setSelectFilter,
      resetSelectFilter,
      setNotSelectFilter
    };
    const tree = renderer
      .create(<FilterSelect {...props} />)
      .toJSON();

      expect(tree).toMatchSnapshot();
  });

  it("render with props", () => {
    const filters = {
      ...initialState.select.elements
    };
    const props = {
      title: "elements",
      filters,
      filterType: "nature",
      setSelectFilter,
      resetSelectFilter,
      setNotSelectFilter
    },
      wrapper = mount(<FilterSelect {...props} />);
    expect(wrapper.props()).toEqual(props);
  });

  it("render with active class", () => {
    const filters = {
      ...initialState.select.elements,
      nature: {
        select: true,
        not: false,
      }
    };
    const props = {
      title: "elements",
      filters,
      filterType: "nature",
      setSelectFilter,
      resetSelectFilter,
      setNotSelectFilter
    },
      wrapper = mount(<FilterSelect {...props} />);
    expect(wrapper.find(".filter-select--value.active").length).toEqual(1);
  });

  it("render with active 'not' class", () => {
    const filters = {
      ...initialState.select.elements,
      nature: {
        select: false,
        not: true,
      }
    };
    const props = {
      title: "elements",
      filters,
      filterType: "nature",
      setSelectFilter,
      resetSelectFilter,
      setNotSelectFilter
    },
      wrapper = mount(<FilterSelect {...props} />);
    expect(wrapper.find(".filter-select--not-value.active").length).toEqual(1);
  });
});
