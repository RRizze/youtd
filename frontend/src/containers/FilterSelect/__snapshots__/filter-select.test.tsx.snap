// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Section tests render correctly 1`] = `
<div
  className="filter-select"
>
  <div
    className="filter-header"
  >
    <h2
      className="filter-title"
    >
      elements
    </h2>
  </div>
  <div
    className="filter-body"
  >
    <ul
      className="filter-select-list"
    >
      <li
        className="filter-select-list__item"
      >
        <span
          className="filter-select-all--value active"
          onClick={[Function]}
        >
          all
        </span>
      </li>
      <li
        className="filter-select-list__item"
      >
        <span
          className="filter-select--value"
          onClick={[Function]}
        >
          nature
        </span>
        <span
          className="filter-select--not-value"
          onClick={[Function]}
        >
          -
        </span>
      </li>
      <li
        className="filter-select-list__item"
      >
        <span
          className="filter-select--value"
          onClick={[Function]}
        >
          fire
        </span>
        <span
          className="filter-select--not-value"
          onClick={[Function]}
        >
          -
        </span>
      </li>
      <li
        className="filter-select-list__item"
      >
        <span
          className="filter-select--value"
          onClick={[Function]}
        >
          ice
        </span>
        <span
          className="filter-select--not-value"
          onClick={[Function]}
        >
          -
        </span>
      </li>
      <li
        className="filter-select-list__item"
      >
        <span
          className="filter-select--value"
          onClick={[Function]}
        >
          astral
        </span>
        <span
          className="filter-select--not-value"
          onClick={[Function]}
        >
          -
        </span>
      </li>
      <li
        className="filter-select-list__item"
      >
        <span
          className="filter-select--value"
          onClick={[Function]}
        >
          darkness
        </span>
        <span
          className="filter-select--not-value"
          onClick={[Function]}
        >
          -
        </span>
      </li>
      <li
        className="filter-select-list__item"
      >
        <span
          className="filter-select--value"
          onClick={[Function]}
        >
          storm
        </span>
        <span
          className="filter-select--not-value"
          onClick={[Function]}
        >
          -
        </span>
      </li>
      <li
        className="filter-select-list__item"
      >
        <span
          className="filter-select--value"
          onClick={[Function]}
        >
          iron
        </span>
        <span
          className="filter-select--not-value"
          onClick={[Function]}
        >
          -
        </span>
      </li>
    </ul>
  </div>
</div>
`;
